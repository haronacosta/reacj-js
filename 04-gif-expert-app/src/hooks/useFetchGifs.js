import { useState, useEffect } from 'react';
import { getGifs } from '../helpers/getGifts';

export const useFetchGifs = (category) => {
  const [state, setState] = useState({
    images: [],
    loading: true,
  });

  useEffect(() => {
    getGifs(category).then((images) => {
      setState({
        images,
        loading: false,
      });
    });
  }, [category]);

  return state;
};
