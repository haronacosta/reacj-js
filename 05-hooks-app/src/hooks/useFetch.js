import { useEffect, useRef, useState } from 'react';

const useFetch = (url) => {
  const isMounted = useRef(true);

  const [state, setState] = useState({
    data: null,
    loading: true,
    error: false,
  });

  useEffect(() => {
    return () => {
      isMounted.current = false;
    };
  }, []);

  useEffect(() => {
    setState({ data: null, loading: true, error: false });

    fetch(url)
      .then((resp) => resp.json())
      .then((data) => {
        if (isMounted.current) {
          setState({ loading: false, error: false, data });
        }
      });
  }, [url]);

  return state;
};

export default useFetch;
