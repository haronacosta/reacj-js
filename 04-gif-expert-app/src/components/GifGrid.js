import React from 'react';
import { useFetchGifs } from '../hooks/useFetchGifs';
import GifGridItem from './GifGridItem';

const GifGrid = ({ category }) => {
  const { loading, images } = useFetchGifs(category);

  return (
    <>
      <h3>{category}</h3>
      {loading ? (
        <h6 className="animate__animated animate__pulse">Cargando... </h6>
      ) : (
        <div className="card-grid">
          {images.map((img) => (
            <GifGridItem key={img.id} {...img} />
          ))}
        </div>
      )}
    </>
  );
};

export default GifGrid;
