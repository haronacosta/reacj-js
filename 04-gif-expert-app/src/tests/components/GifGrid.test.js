import { shallow } from 'enzyme';
import GifGrid from '../../components/GifGrid';
import { useFetchGifs } from '../../hooks/useFetchGifs';
jest.mock('../../hooks/useFetchGifs');

describe('Pruebas en el componente <GifGrid />', () => {
  const category = 'Liverpool';

  test('Debe de regresar correctamente el componente', () => {
    useFetchGifs.mockReturnValue({
      images: [],
      loading: true,
    });

    const wrapper = shallow(<GifGrid category={category} />);

    expect(wrapper).toMatchSnapshot();
  });

  test('Debe mostrar data despues de un gitFetch', () => {
    const gifs = [
      {
        id: 12345,
        title: 'Titulo de prueba',
        url: 'https://localhost.loc/img/unaimagen.jpg',
      },
      {
        id: 1234,
        title: 'Titulo de prueba',
        url: 'https://localhost.loc/img/unaimagen.jpg',
      },
      {
        id: 12445,
        title: 'Titulo de prueba',
        url: 'https://localhost.loc/img/unaimagen.jpg',
      },
    ];

    useFetchGifs.mockReturnValue({
      images: gifs,
      loading: false,
    });

    const wrapper = shallow(<GifGrid category={category} />);

    expect(wrapper).toMatchSnapshot();

    expect(wrapper.find('p').exists()).toBe(false);

    expect(wrapper.find('GifGridItem').length).toBe(gifs.length);
  });
});
