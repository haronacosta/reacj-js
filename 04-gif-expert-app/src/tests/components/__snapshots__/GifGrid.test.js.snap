// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <GifGrid /> Debe de regresar correctamente el componente 1`] = `
<Fragment>
  <h3>
    Liverpool
  </h3>
  <h6
    className="animate__animated animate__pulse"
  >
    Cargando... 
  </h6>
</Fragment>
`;

exports[`Pruebas en el componente <GifGrid /> Debe mostrar data despues de un gitFetch 1`] = `
<Fragment>
  <h3>
    Liverpool
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id={12345}
      key="12345"
      title="Titulo de prueba"
      url="https://localhost.loc/img/unaimagen.jpg"
    />
    <GifGridItem
      id={1234}
      key="1234"
      title="Titulo de prueba"
      url="https://localhost.loc/img/unaimagen.jpg"
    />
    <GifGridItem
      id={12445}
      key="12445"
      title="Titulo de prueba"
      url="https://localhost.loc/img/unaimagen.jpg"
    />
  </div>
</Fragment>
`;
